### ISSUE TITLE
deny transmute `&T` to `&U` where `size_of::<T>` < `size_of::<U>`

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust-clippy#10287]
### What it does

Consider following example:

```rs
let u8_ref: &u8 = &0u8;
let u64_ref: &u64 = unsafe { std::mem::transmute(u8_ref) };
```

This is UB, reported by miri.

Above example can be written:

```rs
let u8_ref: &u8 = &0u8;
let u64_ref: &u64 = &(*u8_ref as u64);
```

Generalize:
let `Original` is original type, and `Dest` is destination type.

If `Original: Clone` and can be coerced into `Dest` by `as`-cast, this can be written as:
```rs
let original_ref: &Original = Original;
let dest_ref: &Dest = &(original_ref.clone() as Dest);
```

If `Original: Clone + Into<Dest>`, this can be written as:
```rs
let original_ref: &Original = Original;
let dest_ref: &Dest = &(original_ref.clone().into());
```

If `Original: AsRef<Dest>`, this can be written as:
```rs
let original_ref: &Original = Original;
let dest_ref: &Dest = original_ref.as_ref();
```

### Lint Name

transmute_to_reference_to_larger_content

### Category

correctness

### Advantage

- Prevents UB.

### Drawbacks

- None.

### Example

```rs
let u8_ref: &u8 = &0u8;
let u64_ref: &u64 = unsafe { std::mem::transmute(u8_ref) };
```

Could be written:

```rs
let u8_ref: &u8 = &0u8;
let u64_ref: &u64 = &(*u8_ref as u64);
```
